### PR TITLE
8333300: [JVMCI] add support for generational ZGC

### PR DESCRIPTION
**Pull request by tkrodriguez**
This exposes the required values for JVMCI to support generational ZGC.  It includes a few things worth mentioning.  JVMCI still exports XBarrierSetRuntime as fields in CompilerToVM::Data under the original name of ZBarrierSetRuntime.  I have exported the XBarrierSetRuntime and ZBarrierSetRuntime functions as addresses under their actual name.  This permits backward compatibility until all the required parts are in place.  We can eventually delete the CompilerToVM::Data names.

I added ZBarrierSetRuntime::load_barrier_on_oop_array paralleling XBarrierSetRuntime::load_barrier_on_oop_array as we use that for a vector barrier.  I could create the function as part of JVMCIRuntime if there are any concerns about including that in the ZGC core.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8333300](https://bugs.openjdk.org/browse/JDK-8333300): [JVMCI] add support for generational ZGC (**Enhancement** - P3)

### Reviewers
 * [Doug Simon](https://openjdk.org/census#dnsimon) (@dougxc - **Reviewer**) ⚠️ Review applies to [8e0cf360](https://git.openjdk.org/jdk/pull/19490/files/8e0cf3605d6f88824fd52f5fd7e4497dc4c66c80)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [8e0cf360](https://git.openjdk.org/jdk/pull/19490/files/8e0cf3605d6f88824fd52f5fd7e4497dc4c66c80)
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to [bb91b42c](https://git.openjdk.org/jdk/pull/19490/files/bb91b42ca4ecf894e2b787a575231853f880560e)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19490/head:pull/19490` \
`$ git checkout pull/19490`

Update a local copy of the PR: \
`$ git checkout pull/19490` \
`$ git pull https://git.openjdk.org/jdk.git pull/19490/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19490`

View PR using the GUI difftool: \
`$ git pr show -t 19490`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19490.diff">https://git.openjdk.org/jdk/pull/19490.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19490#issuecomment-2140831899)

### CODE REVIEW

File: src/hotspot/share/jvmci/jvmci_globals.cpp [hunk 1]
```diff
@@ -223,16 +223,14 @@ bool JVMCIGlobals::enable_jvmci_product_mode(JVMFlagOrigin origin, bool use_graa
...
 }
 
 void JVMCIGlobals::check_jvmci_supported_gc() {
   if (EnableJVMCI) {
     // Check if selected GC is supported by JVMCI and Java compiler
     if (!gc_supports_jvmci()) {
-      log_warning(gc, jvmci)("Setting EnableJVMCI to false as selected GC does not support JVMCI: %s", GCConfig::hs_err_name());
-      FLAG_SET_DEFAULT(EnableJVMCI, false);
-      FLAG_SET_DEFAULT(UseJVMCICompiler, false);
+      fatal("JVMIC does not support the selected GC");
```
Comment (dougxc) [inline src/hotspot/share/jvmci/jvmci_globals.cpp:233]: JVMIC -> JVMCI

**Review by dougxc — APPROVED** [review]

**Review by vnkozlov — APPROVED** [review]
`nmethod.cpp` changes are fine.
ZGC changes have to be reviewed by GC group.

How you tested it to exercise new code (GenZGC + Graal)?

**Comment by vnkozlov** [comment]
And it broke RISC build based on GHA failure for cross compilation.

**Comment by tkrodriguez** [comment]
Testing of the combined Graal and JVMCI changes are in progress.  I hit one main failure that I'm investigating and think I see the problem.  I'll put it the combo all together for testing once I've got that resolved.

**Review by fisk — APPROVED** [review]
Great work. Looks good.

**Comment by tkrodriguez** [comment]
There was one fix I needed to include to use NativeAccess to read from the JVMCI handles when repacking them.  It's only checking for null and not actually accessing the contents so it was benignly broken with singlegen ZGC.  Generational ZGC includes new verify oop logic which was catching this.

**Review by dougxc — APPROVED** [review]

**Comment by dougxc** [comment]
@vnkozlov @fisk any feedback on changes since you approved? I'd love to merge this asap.

**Review by vnkozlov — APPROVED** [review]
Last version is good.

**Comment by dougxc** [comment]
Thanks for all the reviews and input. I'm integrating this on Tom's behalf since he is on vacation.

/integrate

**Comment by fisk** [comment]
Still good.

**Comment by xmas92** [comment]
Just note it still says `JVMIC`.

File: src/hotspot/share/jvmci/jvmci_globals.cpp (hunk 1 above)
Comment (tkrodriguez) [reply src/hotspot/share/jvmci/jvmci_globals.cpp]: I fixed this.

**Comment by tkrodriguez** [comment]
I've tested against latest bits and everything was clean.  Thanks for the reviews.

/integrate

**Comment by DamonFool** [comment]
Hi, please see: https://github.com/openjdk/jdk/pull/19818
Thanks.

**Pull request closed without merging**